In [1]:
import pandas as pd
df = pd.read_csv('Online_Retail.csv')

In [2]:
df.shape

(541909, 8)

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [4]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [5]:
#Raw Revenue Calculation
df['Revenue'] = df['Quantity'] * df['UnitPrice']
raw_revenue = df['Revenue'].sum()
raw_revenue

np.float64(9747747.933999998)

In [6]:
#Cancelled Invoices
cancelled = df[df['InvoiceNo'].astype(str).str.startswith('C')]
cancelled.shape

(9288, 9)

In [7]:
#Negative Quantities- returns
negative_qty = df[df['Quantity'] < 0]
negative_qty.shape

(10624, 9)

In [8]:
#Negative or Zero UnitPrice
invalid_price = df[df['UnitPrice'] <= 0]
invalid_price.shape

#Zero price maybe free item or error, we will exclude

(2517, 9)

In [9]:
#missing CustomerID
missing_customer = df[df['CustomerID'].isnull()]
missing_customer.shape

(135080, 9)

In [10]:
#Duplicate rows
duplicates = df[df.duplicated()]
duplicates.shape

(5268, 9)

In [11]:
df_no_duplicates = df.drop_duplicates()

In [20]:
missing_customer = df[df['CustomerID'].isnull()]
missing_customer.shape

(135080, 9)

In [12]:
df.groupby('Country')['Revenue'].sum().sort_values(ascending=False)

Country
United Kingdom          8187806.364
Netherlands              284661.540
EIRE                     263276.820
Germany                  221698.210
France                   197403.900
Australia                137077.270
Switzerland               56385.350
Spain                     54774.580
Belgium                   40910.960
Sweden                    36595.910
Japan                     35340.620
Norway                    35163.460
Portugal                  29367.020
Finland                   22326.740
Channel Islands           20086.290
Denmark                   18768.140
Italy                     16890.510
Cyprus                    12946.290
Austria                   10154.320
Hong Kong                 10117.040
Singapore                  9120.390
Israel                     7907.820
Poland                     7213.140
Unspecified                4749.790
Greece                     4710.520
Iceland                    4310.000
Canada                     3666.380
Malta               

In [ ]:
clean_df = df.copy()

clean_df = clean_df[~clean_df['InvoiceNo'].astype(str).str.startswith('C')]
clean_df = clean_df[clean_df['Quantity'] > 0]
clean_df = clean_df[clean_df['UnitPrice'] > 0]
clean_df = clean_df.drop_duplicates()

clean_df.shape

(524878, 9)

In [14]:
clean_df['Revenue'] = clean_df['Quantity'] * clean_df['UnitPrice']
clean_revenue = clean_df['Revenue'].sum()
clean_revenue

np.float64(10642110.804)

In [15]:
raw_revenue, clean_revenue, raw_revenue - clean_revenue

(np.float64(9747747.933999998),
 np.float64(10642110.804),
 np.float64(-894362.870000001))

clean revenue is more because of removal of -ve unit price, -ve quantity(returns) and cancelled orders

In [16]:
clean_df.groupby('Description')['Quantity'].sum().sort_values(ascending=False).head()

#Best-selling products should be prioritized in inventory

Description
PAPER CRAFT , LITTLE BIRDIE           80995
MEDIUM CERAMIC TOP STORAGE JAR        78033
WORLD WAR 2 GLIDERS ASSTD DESIGNS     54951
JUMBO BAG RED RETROSPOT               48371
WHITE HANGING HEART T-LIGHT HOLDER    37872
Name: Quantity, dtype: int64

In [17]:
clean_df.groupby('Country')['Revenue'].sum().sort_values(ascending=False)

#UK dominates revenue, so it should be the core market focus

Country
United Kingdom          9001744.094
Netherlands              285446.340
EIRE                     283140.520
Germany                  228678.400
France                   209625.370
Australia                138453.810
Spain                     61558.560
Switzerland               57067.600
Belgium                   41196.340
Sweden                    38367.830
Japan                     37416.370
Norway                    36165.440
Portugal                  33683.050
Finland                   22546.080
Singapore                 21279.290
Channel Islands           20440.540
Denmark                   18955.340
Italy                     17483.240
Hong Kong                 15483.000
Cyprus                    13502.850
Austria                   10198.680
Israel                     8129.410
Poland                     7334.650
Greece                     4760.520
Unspecified                4740.940
Iceland                    4310.000
Canada                     3666.380
USA                 

In [18]:
returns_df = df[df['Quantity'] < 0].copy()
returns_df['ReturnValue'] = returns_df['Quantity'] * returns_df['UnitPrice']

total_returns = returns_df['ReturnValue'].sum()
total_returns

np.float64(-896812.49)

  - Make CustomerID mandatory
  - fix the platform as there are 5,268 duplicates
  - launch a dedicated loyalty or VIP program for top-tier UK buyers to guard against domestic competitors as UK is top revenue performer and work on marketing in low performing countries
  - Best-selling products should be prioritized 